### DISCUSSION TITLE
The search method of the p-autocomplete component has been changed

### DISCUSSION THREAD

**Discussion by mlc-mlapis** [primefaces/community#510] [PrimeNG] [answered]
The API of the `search` method has changed since PrimeNG 16.7.0. It is a public method of the `p-autocomplete` component and is valuable in many cases. Why it is not mentioned in the latest documentation needs to be clarified. It would be beneficial to understand the actual required API, which is:

`AutoComplete.search(event: any, query: any, source: any): void`

The change is that a third new parameter has been added, and what it means or which values are allowed needs to be clarified.

Thank you so much for your attention and participation.

**Comment by cetincakiroglu** [comment]
Hi,

With version 16.7.0, we have released the first part of the accessibility implementation for inputs that we have been working on for a long time. Based on the research we conducted during the implementation, we decided that many components needed changes in their structure and simplification, and these components were rewritten. Because of this, search method is also rewritten and since it's not publicly shown method. Therefore, we have also rewritten and updated the search method. This method was not previously mentioned in the documentation, no examples were given for users to use it, and since it's a method used solely internally by the component, there's no need for us to mention it anywhere. While this method is not a prop, its usage is at your discretion. Additionally, you can review the changes we made for input components in the 16.7.2 version changelog.

**Reply by mlc-mlapis** [reply]
@cetincakiroglu, I understand what you're explaining. However, the `search` method is logical when you want to programmatically control and use the `p-autocomplete` component (taking its instance via the @ViewChild decorator). Can you suggest any equivalent way to get the same result? How would you use it except for a manual direct control by a user?

It would benefit everybody if such a method were documented and available for general use.

**Reply by cetincakiroglu** [reply]
You can still access the search function, it's refactored but I assume that you can still use that. However, allowing users to create and use their own search function is a good idea. We may add a new input property that allows users to customize their search. We may add that as a feature in v17. I liked the idea.

New search function after the refactor is as follows;
```
    search(event, query, source) {
        //allow empty string but not undefined or null
        if (query === undefined || query === null) {
            return;
        }

        //do not search blank values on input change
        if (source === 'input' && query.trim().length === 0) {
            return;
        }
        this.loading = true;
        this.completeMethod.emit({ originalEvent: event, query });
    }
```

**Reply by cetincakiroglu** [reply]
Could you please try and give feedback if you can use it? If it still works, we can consider to make it public and put an example about "Custom search implementation"

**Reply by mlc-mlapis** [reply]
Yep, we will use it, of course. 😄 I wanted to be sure I understood the changed logic and your reasons. I would appreciate it if you added it in v17.

**Reply by cetincakiroglu** [reply, marked as answer]
Yes it would be good for users, I appreciate for the idea :)

**Reply by mlc-mlapis** [reply]
I'll do some tests and report to you as soon as possible.

**Reply by mlc-mlapis** [reply]
@cetincakiroglu, I have looked at the source code and tried to use some cases. Using the actual version of the `search` method would be fine. The only thing is probably some other variations of the third `source` parameter used in the `p-autocomplete` component. They are the following:
```
// The only one that should be documented, actually.
onInput -> search(event, query, 'input')
// The value 'focus' is not directly used in the search code, only as an `else` path.
onInputFocus -> search(event, query, 'focus')
// The value 'dropdown' is not directly used in the search code, only as an `else` path.
handleDropdownClick -> search(event, query, 'dropdown')
```
Are there any ideas on modifying the search method and directly using the `focus` and `dropdown` variants?

**Reply by cetincakiroglu** [reply]
I guess we need to take a callback function from users as we do in table filtering. Would be nice new feature. Could you please create stackblitz examples for the cases you've tried, it would be helpful to see what is possible and what is not.

**Reply by mlc-mlapis** [reply]
@cetincakiroglu So, I reviewed our used actual cases, and as a summary, here it is:
* allow writing a searched string without focusing the input with autocomplete (using RxJS `fromEvent`) and programmatically run the `search`  method,
* use arrows on the keyboard (↑ ↓ → ←) to control what and how it is searched through autocomplete functionality concerning the history or the context.

All of the cases relate to the comfort and speed of how a user can control applications directly from a keyboard without a mouse and are widely spread through many places in applications.
